src/Lean/Data/Trie.lean

```diff
@@ -1,110 +1,202 @@
 /-
 Copyright (c) 2018 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Author: Sebastian Ullrich, Leonardo de Moura
+Author: Sebastian Ullrich, Leonardo de Moura, Joachim Breitner
 
-Trie for tokenizing the Lean language
+A string trie data strucuture, used for tokenizing the Lean language
 -/
 import Lean.Data.Format
 
 namespace Lean
-namespace Parser
+namespace Data
 
+/-
+## Implementation notes
+
+Tries have typically many nodes with small degree, where a linear scan
+through the (compact) `ByteArray` is faster than using binary search or
+search trees like `RBTree`.
+
+Moreover, many nodes have degree 1, which justifies the special case `Node1`
+constructor.
+
+The code would be a bit less repetitive if we used something like the following
+```
+mutual
+def Trie α := (Option α, ByteAssoc α)
+
+inductive ByteAssoc α where
+  | Leaf : Trie α
+  | Node1 : UInt8 → Trie α → Trie α
+  | Node : ByteArray -> Array (Trie α) → Trie α
+end
+```
+but that would come at the cost of extra indirections.
+-/
+
+/-- A Trie is a key-value store where the keys are of type `String`,
+and the internal structure is a tree that branches on the bytes of the string.  -/
 inductive Trie (α : Type) where
-  | Node : Option α → RBNode Char (fun _ => Trie α) → Trie α
+  | Leaf : Option α → Trie α
+  | Node1 : Option α → UInt8 → Trie α → Trie α
+  | Node : Option α → ByteArray -> Array (Trie α) → Trie α
```

Comment:
Oops, all these constructors should be/have been lower-cased. Also there's a stray ASCII arrow.